### ISSUE TITLE
Add SWC Plugin

### ISSUE THREAD

**Issue by heygrady** [kentcdodds/babel-plugin-macros#144]
https://swc-project.github.io/docs/usage-plugin

**Proposal:** create an `swc-plugin-macros` that is similar to `babel-plugin-macros`

[SWC](https://github.com/swc-project/swc) is increasing in popularity. It is aiming to be a drop-in replacement for babel with some significant [speed improvements](https://swc-project.github.io/blog/2020/02/16/why-is-swc-fast). It already has a [webpack loader](https://swc-project.github.io/docs/usage-swc-loader).

There is some early interest in how it might integrate with create-react-app (https://github.com/facebook/create-react-app/issues/8060). Looking at [which babel plugins create-react-app uses](https://github.com/facebook/create-react-app/blob/master/packages/babel-preset-react-app/package.json#L25-L38) it seems like most of those things are [supported by swc](https://swc-project.github.io/docs/comparison-babel) (with only a few exceptions).

SWC appears to be missing support for:

1. **`babel-plugin-macros`** (important for some of my projects)
2. `babel-plugin-transform-react-remove-prop-types` (less important in TypeScript projects)
3. `babel-plugin-transform-flow-strip-types` (generally SWC has no Flow support 🤷🏻‍♂️)

Generally everything else babel does for me is handled by [swc preset-env](https://swc-project.github.io/docs/preset-env) (or isn't particularly vital to my projects).

**Comment by kentcdodds** [comment]
Hi @heygrady!

That sounds cool. I don't have the bandwidth to work on it myself, but if you want to do it then you're welcome to! 👍

**Comment by c-rick** [comment]
try this code, it work for me
[next-twin](https://github.com/c-rick/next-twin)

**Comment by conartist6** [comment]
What language are SWC plugins written in? Rust?

**Comment by conartist6** [comment]
Indeed, they are written in Rust. So macros would have to be themselves written in Rust. This may be possible, but it won't share any code or architecture with `babel-plugin-macros` and would require maintainers with a completely different skill set. I'm going to close this issue as it will never be actionable here. Best of luck to those looking for this functionality.

**Comment by renchap** [comment]
Some update on the SWC front here: https://github.com/swc-project/swc/issues/3167

TL;DR: they are integrating WASM plugin support in SWC. This means you can write your plugins in any language that can be compiled to WASM, for example Typescript with https://www.assemblyscript.org

Would this make it easier to have a macros plugin for SWC ?

**Comment by conartist6** [comment]
That might make it possible! May I suggest that you make a repo or chat or something for people who are interested in using or building that functionality to talk and collaborate, and you should link to it here.

**Comment by conartist6** [comment]
I agree that it would be a separate project that would not be intercompatible with this one. Hence the suggestion to make a new space for interested contributors.

**Comment by matthewmatician** [comment]
Maybe this is a noob question, but wouldn't [this PR](https://github.com/swc-project/swc/pull/1465) have made this possible?

**Comment by conartist6** [comment]
@matthewmatician I mean you could always do a two-step process. Parse with babel, transpile macros, send the program to SWC, then have SWC do the rest of the transpilation.

The only question is: how you would send the program to SWC? Obviously you could print it from babel and parse in SWC. But perhaps there's a more efficient way? I can't say I understand the PR you linked well enough to know quite what its API would be. If it can do a copy in a way that's faster than a parse and print, then perhaps the perf numbers would still be favorable at the end of the day.

**Comment by arlyon** [comment]
I am working on [stailwc](https://github.com/arlyon/stailwc), a swc-native version of [twin.macro](https://github.com/ben-rogerson/twin.macro) which serves as both
a) a neat project (my build times are down ~80% compared to bazel!)
b) a living example on how to build an swc plugin for nextjs
